daemon/mgr/spec_linux.go

```diff
@@ -100,41 +118,487 @@ func setupSeccomp(ctx context.Context, meta *ContainerMeta, spec *SpecWrapper) e
+		return err
+	}
+
+	s.Linux.Resources.BlockIO = &specs.LinuxBlockIO{
+		Weight:                  &r.BlkioWeight,
+		WeightDevice:            weightDevice,
+		ThrottleReadBpsDevice:   readBpsDevice,
+		ThrottleReadIOPSDevice:  readIOpsDevice,
+		ThrottleWriteBpsDevice:  writeBpsDevice,
+		ThrottleWriteIOPSDevice: writeIOpsDevice,
+	}
+
+	return nil
+}
+
+func getWeightDevice(devs []*types.WeightDevice) ([]specs.LinuxWeightDevice, error) {
+	var stat syscall.Stat_t
+	var weightDevice []specs.LinuxWeightDevice
+
+	for _, dev := range devs {
+		if err := syscall.Stat(dev.Path, &stat); err != nil {
+			return nil, err
+		}
+
+		d := specs.LinuxWeightDevice{
+			Weight: &dev.Weight,
+		}
+		d.Major = int64(stat.Rdev / 256)
+		d.Minor = int64(stat.Rdev % 256)
+		weightDevice = append(weightDevice, d)
+	}
+
+	return weightDevice, nil
+}
+
+func getThrottleDevice(devs []*types.ThrottleDevice) ([]specs.LinuxThrottleDevice, error) {
+	var stat syscall.Stat_t
+	var ThrottleDevice []specs.LinuxThrottleDevice
+
+	for _, dev := range devs {
+		if err := syscall.Stat(dev.Path, &stat); err != nil {
+			return nil, err
+		}
+
+		d := specs.LinuxThrottleDevice{
+			Rate: dev.Rate,
+		}
+		d.Major = int64(stat.Rdev / 256)
+		d.Minor = int64(stat.Rdev % 256)
+		ThrottleDevice = append(ThrottleDevice, d)
+	}
+
+	return ThrottleDevice, nil
+}
+
+// setupResource creates linux cpu resource spec
+func setupCPU(ctx context.Context, r types.Resources, s *specs.Spec) {
+	cpu := &specs.LinuxCPU{
+		Cpus: r.CpusetCpus,
+		Mems: r.CpusetMems,
+	}
+
+	if r.CPUShares != 0 {
+		v := uint64(r.CPUShares)
+		cpu.Shares = &v
+	}
+
+	if r.CPUPeriod != 0 {
+		v := uint64(r.CPUPeriod)
+		cpu.Period = &v
+	}
+
+	if r.CPUQuota != 0 {
+		v := int64(r.CPUQuota)
+		cpu.Quota = &v
+	}
+
+	s.Linux.Resources.CPU = cpu
+}
+
+// setupResource creates linux memory resource spec.
+func setupMemory(ctx context.Context, r types.Resources, s *specs.Spec) {
+	memory := &specs.LinuxMemory{}
+	if r.Memory != 0 {
+		v := r.Memory
+		memory.Limit = &v
+	}
+
+	if r.MemorySwap != 0 {
+		v := r.MemorySwap
+		memory.Swap = &v
+	}
+
+	if r.MemorySwappiness != nil {
+		v := uint64(*r.MemorySwappiness)
+		memory.Swappiness = &v
+	}
+
+	if r.OomKillDisable != nil {
+		v := bool(*r.OomKillDisable)
+		memory.DisableOOMKiller = &v
+	}
+
+	s.Linux.Resources.Memory = memory
+}
+
+// setupResource creates linux device resource spec.
+func setupDevices(ctx context.Context, c *ContainerMeta, s *specs.Spec) error {
+	var devs []specs.LinuxDevice
+	devPermissions := s.Linux.Resources.Devices
+	if c.HostConfig.Privileged {
+		hostDevices, err := devices.HostDevices()
+		if err != nil {
+			return err
+		}
+		for _, d := range hostDevices {
+			devs = append(devs, linuxDevice(d))
+		}
+		devPermissions = []specs.LinuxDeviceCgroup{
+			{
+				Allow:  true,
+				Access: "rwm",
+			},
+		}
+	} else {
+		for _, deviceMapping := range c.HostConfig.Devices {
+			if !opts.ValidateDeviceMode(deviceMapping.CgroupPermissions) {
```

Comment:
I am wondering if we have to do the validation checking here. Is there any possibility to put this ahead to validations in manager layer instead of spec layer.